odoo/addons/test_new_api/tests/test_properties.py

```diff
@@ -0,0 +1,1064 @@
+            'message_properties': message_properties_1,
+            'participants': [Command.link(cls.user.id)],
+        })
+        cls.discussion_2 = cls.env['test_new_api.discussion'].create({
+            'name': 'Test Discussion',
+            'message_properties': message_properties_2,
+            'participants': [Command.link(cls.user.id)],
+        })
+
+        cls.message_1 = cls.env['test_new_api.message'].create({
+            'name': 'Test Message',
+            'discussion': cls.discussion_1.id,
+            'author': cls.user.id,
+            'custom_properties': {'name': 'Test', 'partner_id': [cls.partner.id, 'res.partner']},
+        })
+
+        cls.env.flush_all()
+
+        cls.message_2 = cls.env['test_new_api.message'].create({
+            'name': 'Test Message',
+            'discussion': cls.discussion_1.id,
+            'author': cls.user.id,
+        })
+        cls.message_3 = cls.env['test_new_api.message'].create({
+            'name': 'Test Message',
+            'discussion': cls.discussion_2.id,
+            'author': cls.user.id,
+        })
+
+    def test_json_field(self):
+        author = self.env['res.users'].search([], limit=1)
+
+        json_values = [
+            {
+                'id': 'name',
+                'string': 'Name',
+                'type': 'char',
+            }, {
+                'id': 'partner_id',
+                'string': 'Partner',
+                'type': 'many2one',
+                'model': 'res.partner',
+            },
+        ]
+
+        discussion = self.env['test_new_api.discussion'].create({
+            'name': 'Test Discussion',
+            'json_values': json_values,
+            'participants': [Command.link(author.id)],
+        })
+
+        self._flush_and_invalidate()
+
+        self.assertTrue(isinstance(discussion.json_values, list))
+        self.assertEqual(discussion.json_values, json_values)
+
+        discussion.json_values = '{"test": 3}'
+        self.assertEqual(
+            discussion.json_values, {"test": 3},
+            msg='JSON must be parsed when loaded from cache')
+
+        discussion.invalidate_recordset()
+
+        self.assertEqual(discussion.json_values, {"test": 3})
+
+    def test_properties_field(self):
+        message_1, message_2, message_3 = self.message_1, self.message_2, self.message_3
+        partner = self.partner
+
+        self._flush_and_invalidate()
+
+        self.assertTrue(isinstance(message_1.custom_properties, list))
+        self.assertEqual(message_1.custom_properties[0]['value'], 'Test')
+        self.assertEqual(message_1.custom_properties[1]['value'], partner.id)
+        self.assertEqual(message_2.custom_properties[0]['value'], 'Default Name')
+        self.assertFalse(message_2.custom_properties[1]['value'])
+        self.assertEqual(
+            message_3.custom_properties[0]['value'], 'draft',
+            msg='Should have take the default value')
+
+        message_1.custom_properties = {
```

Comment:
So when you read it's an array but when you write you write a dict?
This seems a bit strange.

Goes back to my previous comment, maybe it should always be a dict?
The structure also seems different.

Could be annoying because usually in the code you want to be able to search a record, then read the properties values and receive a particular structure, then write to change it, then read again and expect to receive the same kind of structure.
But it seems like it would not be the case here?

(This is maybe not clear, don't hesitate to ask on Discord if so).